data/builder/file.go

```diff
@@ -0,0 +1,198 @@
+package builder
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/ipfs/go-cid"
+	chunk "github.com/ipfs/go-ipfs-chunker"
+	"github.com/ipfs/go-unixfsnode/data"
+	dagpb "github.com/ipld/go-codec-dagpb"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	basicnode "github.com/ipld/go-ipld-prime/node/basic"
+	"github.com/multiformats/go-multicodec"
+	multihash "github.com/multiformats/go-multihash/core"
+)
+
+// BuildUnixFSFile creates a dag of ipld Nodes representing file data.
+// This recreates the functionality previously found in
+// github.com/ipfs/go-unixfs/importer/balanced, but tailored to the
+// go-unixfsnode & ipld-prime data layout of nodes.
+// We make some assumptions in building files with this builder to reduce
+// complexity, namely:
+// * we assume we are using CIDv1, which has implied that the leaf
+//   data nodes are stored as raw bytes.
+//   ref: https://github.com/ipfs/go-mfs/blob/1b1fd06cff048caabeddb02d4dbf22d2274c7971/file.go#L50
+func BuildUnixFSFile(r io.Reader, chunker string, ls *ipld.LinkSystem) (ipld.Link, error) {
```

Comment:
Do you think there's also a way we could make a `NodeAssembler` type somewhere that has a constructor taking a `chunker string, ls *ipld.LinkSystem`, and then we can have its `AssignBytes` method call this?

It wouldn't be memory-efficient of course, but the fact that it works at all would be nice.

More importantly, it would get us within a centimeter of adding an `AssignFromReader` method to that type, or something of that nature, which we could later let people feature-detect for.

Reply:
does it make sense to do it at that level, or would we prefer the assembler a level up where it could handle either file or directory assembly (switching on being either given bytes or a map listing of nodes)

Reply:
(Still kinda wanna do something about this but don't have an exact fullbaked idea of how I'd want it to look, either, so it's definitely not something to hold up the rest of this PR landing on.)